Documentation/Compilation-Errors.md

```diff
@@ -0,0 +1,134 @@
+# Compilation errors
+
+Swift’s compile errors can be mystifying if you are not intimately familiar with
+the internals of Argo. Here we describe some of the most common errors and how
+to fix them.
+
+## Complex expressions
+
+Considering that the implementation of `decode` on a model consists of fully
+currying a method and composing it with many small pieces, each of which may be
+the composition of pieces, it is not surprising that Swift sometimes has
+difficulty compiling the expression in a reasonable amount of time. This
+manifests itself with the following error:
+
+> Expression was too complex to be solved in reasonable time; consider
+> breaking up the expression into distinct sub-expressions.
+
+Fortunately this is easy to solve. A first step is to store
+`curry(Model.init)` in a variable instead of using it directly:
+
+```swift
+static func decode(j: JSON) -> Decoded<Model> {
+  let create = Curry(Model.init)
+  return create
+    <^> j <| "key1"
+    <*> j <| "key2"
+    ...
+```
+
+Many times this can fix the problem. If you still have complex expression
+issues, the next step is to pick a midpoint in your decoding chain and set an
+intermediate variable:
+
+```swift
+static func decode(j: JSON) -> Decoded<Model> {
+  let create = Curry(Model.init)
+  let tmp = create
+    <^> j <| "key_a_1"
+    <*> j <| "key_a_2"
+    ...
+  return tmp
+    <*> j <| "key_b_1"
+    <*> j <| "key_b_2"
+    ...
+```
+
+This should almost surely fix all of your complex expression issues, but if not,
```

Comment:
"should almost surely" sounds alternately very confident and very uncertain. How about: "This will almost always fix...".